src/main/scala/org/constellation/p2p/PeerAPI.scala

```diff
@@ -143,43 +145,109 @@ class PeerAPI(override val ipManager: IPManager[IO])(
 
   private[p2p] def postEndpoints(socketAddress: InetSocketAddress) =
     post {
-      pathPrefix("snapshot") {
-        path("info") {
-          APIDirective.extractIP(socketAddress) { ip =>
-            entity(as[Array[Array[Byte]]]) { curCheckpointHashes =>
+      pathPrefix("snapshot" / "info") {
+        pathEnd {
+          entity(as[(String, Array[Array[Byte]])]) {
+            case (address, curCheckpointHashes) =>
               val deSerCheckpointHashes =
                 curCheckpointHashes.flatMap(chunkDeSerialize[Seq[String]](_, "snapshot/info/curCheckpointHashes"))
               val getInfo = dao.snapshotService.getSnapshotInfo.flatMap { info =>
-                logger.warn(s"snapshot/info info.acceptedCBSinceSnapshot.size - ${info.acceptedCBSinceSnapshot}")
-                logger.warn(s"snapshot/info curCheckpointHashes.size - ${deSerCheckpointHashes.size}")
-
-                val checkpointsToGet = info.acceptedCBSinceSnapshot.toList.diff(deSerCheckpointHashes.toList)
-
-                logger.warn(s"snapshot/info checkpointsToGet.size - ${checkpointsToGet.size}")
-                logger.warn(s"snapshot/info checkpointsToGet - ${checkpointsToGet}")
-
+                val checkpointsToGet = info.acceptedCBSinceSnapshotHashes.toList.diff(deSerCheckpointHashes.toList)
                 checkpointsToGet.traverse {
                   dao.checkpointService.fullData(_)
                 }.flatMap { cbs =>
                   val infoSer = info.copy(acceptedCBSinceSnapshotCache = cbs.flatten).toSnapshotInfoSer()
-                  dao.snapshotService.recentSnapshotInfo
-                    .put(ip, infoSer)
+                  val res = dao.snapshotService.recentSnapshotInfo
+                    .put(address, infoSer)
                     .flatTap(_ => IO { logger.warn(s"snapshot/info recentSnapshotInfo updating") })
                     .map(_ => Array.empty[Byte].some)
+                  res
                 }
               }
-
               APIDirective.handle {
                 dao.cluster.getNodeState
                   .map(NodeState.canActAsRedownloadSource)
                   .ifM(getInfo, IO.pure(None))
               } {
                 case None => complete(StatusCodes.NotFound)
-                case _    => complete(StatusCodes.OK, Array.empty[Byte])
+                case _    => complete(StatusCodes.OK, Array()) //need Array() here, Array.empty marshalls as List
               }
+          }
+        } ~
+          path(LongNumber) { height =>
+            entity(as[(String, Array[Array[Byte]])]) {
+              case (address, curCheckpointHashes) =>
+                val curSnap = dao.snapshotService.storedSnapshot.get.unsafeRunSync()
```

Comment:
Get rid of `unsafeRunSync`. It will lead to performance issues (nested unsafeRunSync)

Reply:
So this is where I was mentioning today in daily. First, we can handle both cases in by just iterating over dao.snapshotService.storedSnapshot.get and making this a Get once again. We aught to just send curCheckpointHashes to "snapshot" / "obj" / "acceptedCBSinceSnapshotCache" to then filter, then reSerialize and send over. I wouldn't worry about the double serialization for now